### PR TITLE
[RFC] Add assertEventFired and assertEventFiredWith

### PR DESCRIPTION
**Pull request by jeremyharris**
This is an RFC for a compliment to #8853 that allows developers to assert that events were fired, or fired with data. Spawned from a personal project, the addition of event lists makes this more of a candidate for the core than it was before.

As a practical example, I've added it to TableTest::testAfterSave below:

``` php
    /**
     * Asserts that afterSave callback is called on successful save
     *
     * @group save
     * @return void
     */
    public function testAfterSave()
    {
        $table = TableRegistry::get('users');
        $data = new \Cake\ORM\Entity([
            'username' => 'superuser',
            'created' => new Time('2013-10-10 00:00'),
            'updated' => new Time('2013-10-10 00:00')
        ]);

        $table->eventManager()->setEventList(new \Cake\Event\EventList());
        $table->eventManager()->trackEvents(true);

        $called = false;
        $listener = function ($e, $entity, $options) use ($data, &$called) {
            $this->assertSame($data, $entity);
            $this->assertTrue($entity->dirty());
            $called = true;
        };
        $table->eventManager()->on('Model.afterSave', $listener);

        $calledAfterCommit = false;
        $listenerAfterCommit = function ($e, $entity, $options) use ($data, &$calledAfterCommit) {
            $this->assertSame($data, $entity);
            $this->assertFalse($entity->dirty());
            $calledAfterCommit = true;
        };
        $table->eventManager()->on('Model.afterSaveCommit', $listenerAfterCommit);

        $this->assertSame($data, $table->save($data));
        $this->assertEquals($data->id, self::$nextUserId);
        $this->assertTrue($called);
        $this->assertTrue($calledAfterCommit);

        $this->assertEventFired('Model.afterSave', $table->eventManager());
    }
```

By default, it the assertions use the global event manager, so asserting global events were fired would look something like this:

```
$this->assertEventFired('My.Global.Event');
$this->assertEventFiredWith('My.Global.Event', 'data', 'value');
```

The downside is that it's only useable if track events is on, so perhaps it would better be in a trait that the user has to opt-in. I was going to put this into a plugin but it seemed a bit silly so I thought I'd suggest it here first, if it doesn't seem like a feature that would be useful/good for the core I can pull it out into a plugin when event lists are released.

### CODE REVIEW

File: src/TestSuite/Constraint/EventFired.php [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+    /**
+     * Checks if event is in fired array
+     *
+     * @param mixed $other Constraint check
+     * @return bool
+     */
+    public function matches($other)
+    {
+        return $this->_eventManager->getEventList()->hasEvent($other);
```
Comment (markstory) [inline src/TestSuite/Constraint/EventFired.php]: What if `getEventList()` returns null?

File: src/TestSuite/Constraint/EventFiredWith.php [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+     * @param mixed $other Constraint check
+     * @return bool
+     */
+    public function matches($other)
+    {
+        $firedEvents = [];
+        $list = $this->_eventManager->getEventList();
+        $totalEvents = count($list);
+        for ($e = 0; $e < $totalEvents; $e++) {
+            $firedEvents[] = $list[$e];
```
Comment (markstory) [inline src/TestSuite/Constraint/EventFiredWith.php]: Perhaps `EventList` should be an Iterator?

Comment (jeremyharris) [inline src/TestSuite/Constraint/EventFiredWith.php]: Yeah this was uggggglly. I didn't modify anything in event list but it would definitely help.

File: src/TestSuite/Constraint/EventFired.php (hunk 1 above)
Comment (jeremyharris) [inline src/TestSuite/Constraint/EventFired.php]: Was waiting on some general yes/no before handling all cases haha. But yes, both constraints would check that we were tracking events and that there was an event list, if not it would throw an exception. (probably in __construct).

**Comment by lorenzo** [comment]
I like the idea

File: src/TestSuite/Constraint/EventFiredWith.php (hunk 2 above)
Comment (markstory) [inline src/TestSuite/Constraint/EventFiredWith.php]: Could be a refinement for a future change.

Comment (jeremyharris) [inline src/TestSuite/Constraint/EventFiredWith.php]: Should I submit a separate PR for that change or do you mind me throwing it in here? I don't mind either way, although if it's separate this PR should wait for it otherwise there'll be another one to update this block :)

**Comment by jeremyharris** [comment]
I will update this tomorrow to handle missing event lists and managers that aren't tracking events.

Also, I didn't see how event lists are instantiated in the core. Is it up to the user? If this gets merged I just want to know what documentation to add :)

File: src/TestSuite/Constraint/EventFiredWith.php [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+     */
+    public function __construct($eventManager, $dataKey, $dataValue)
+    {
+        parent::__construct();
+        $this->_eventManager = $eventManager;
+        $this->_dataKey = $dataKey;
+        $this->_dataValue = $dataValue;
+
+        if ($this->_eventManager->getEventList() === null) {
+            throw new PHPUnit_Framework_AssertionFailedError('The event manager you are asserting against is not configured to track events.');
```
Comment (jeremyharris) [inline src/TestSuite/Constraint/EventFiredWith.php]: There's no current way to check the current state of `trackEvents` so I left it off. Setting an event list automatically sets them to be tracked, so it seems the developer needs to deliberately `trackEvents(false)` for it to fail, in which case it just won't show in the list and the assertion would fail anyway. Adding the ability to check the track status could be part of improving EventList in a separate PR, which I'm happy to do.

Comment (markstory) [inline src/TestSuite/Constraint/EventFiredWith.php]: I think checking `getEventList()` is good enough.

File: src/TestSuite/Constraint/EventFiredWith.php (hunk 2 above)
Comment (markstory) [inline src/TestSuite/Constraint/EventFiredWith.php]: I think it can be separate.

**Comment by jeremyharris** [comment]
Build broke but I can't restart it.

Also, I've got a commit with some documentation on event lists and testing events in the works here: https://github.com/jeremyharris/docs/commit/d122fcc6a45188d88b66a569fbd981bc0c964d23. (It assumes this will be merged and released in `.10` but I can correct those assumptions at any time.)

**Comment by markstory** [comment]
Looks good to me.

**Comment by markstory** [comment]
@jeremyharris Nice work! Looking forward tohe docs 😀
